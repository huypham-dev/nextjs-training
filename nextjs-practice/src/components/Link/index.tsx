import { memo } from "react";

import Link, { LinkProps } from "next/link";

interface Props extends LinkProps {
  children: React.ReactNode;
  className?: string;
}

const LinkBase = ({ children, className, ...linkProps }: Props) => (
  <Link {...linkProps}>
    <a className={className}>{children}</a>
  </Link>
);

export default memo(LinkBase);
