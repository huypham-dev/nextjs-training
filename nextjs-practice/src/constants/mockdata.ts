export const CATEGORIES = [
  {
    id: "1",
    name: "mobile",
    image:
      "https://images.prismic.io/nextjsexample/7853f863-6a12-4312-a6da-f930437f6d62_mobile.png",
  },
  {
    id: "2",
    name: "internet",
    image:
      "https://images.prismic.io/nextjsexample/cb71ce70-bb03-442b-99e8-a318255b81f3_3579187.png",
  },
];

export const ARTICLES = [
  {
    id: "1",
    createdAt: "2021-03-01T12:59:16.855Z",
    title: "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
    author: "author 1",
    image:
      "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
    category: "internet",
    content:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
    slug: "slug 1",
    view: 90,
  },
  {
    id: "2",
    createdAt: "2021-03-01T12:59:16.855Z",
    title: "Lorem ipsum dolor sit amet, consectetur adipisicing elit",
    author: "author 1",
    image:
      "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg",
    category: "internet",
    content:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.1",
    slug: "slug 1",
    view: 90,
  },
];
