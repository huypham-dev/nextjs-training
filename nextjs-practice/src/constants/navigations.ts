export const navigations = [
  { label: "Home", path: "/" },
  { label: "Blog", path: "/blog" },
  { label: "Contact", path: "/contact" },
];
