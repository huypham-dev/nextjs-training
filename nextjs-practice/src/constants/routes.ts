export const ROUTES = {
  home: {
    path: "/",
  },
  blog: {
    path: "/blog",
    withId: (id: string) => `/blog/${id}`,
  },
  post: {
    path: "/post",
    withId: (id: string) => `/post/${id}`,
  },
  contact: {
    path: "/",
  },
};
