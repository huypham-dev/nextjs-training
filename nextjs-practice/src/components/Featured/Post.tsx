import { memo } from "react";

import { PostType } from "types";
import LinkBase from "components/Link";
import Image from "next/image";
import isEqual from "react-fast-compare";
import { ROUTES } from "constants/routes";
import Typography from "components/common/Typography";

type Props = {
  post: PostType;
};

const Post = ({ post }: Props) => {
  const { id, title, image, view } = post;

  return (
    <div className="w-full md:w-1/2 lg:w-1/3 px-6 p-5">
      <div className="bg- p-4 rounded-md shadow-xl border-gray-200 border-2">
        <Typography
          variant="h2"
          className="text-md font-extrabold tracking-widest mb-2"
        >
          {title}
        </Typography>
        <div className="w-full text-center">
          <Image
            className="rounded-lg"
            src={image}
            alt="article image"
            width={250}
            height={150}
            layout="responsive"
          />
        </div>
        <Typography className="text-xs text-gray-500 mb-4">
          {view} <i>views</i>
        </Typography>
        <LinkBase
          href={ROUTES.post.withId(id)}
          className="inline-block border-2 border-black py-3 px-8 uppercase font-display font-thin tracking-wide text-xs hover:bg-yellow-400 hover:border-yellow-400"
        >
          Read more
        </LinkBase>
      </div>
    </div>
  );
};

export default memo(Post, isEqual);
