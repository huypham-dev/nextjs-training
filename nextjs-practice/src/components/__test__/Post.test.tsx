import { mount, shallow } from "enzyme";
import Link from "next/link";
import Image from "next/image";
import Post from "../Featured/Post";
import { ARTICLES } from "constants/mockData";

describe("Post", () => {
  it("Snapshot", () => {
    const component = shallow(<Post post={ARTICLES[0]} />);

    expect(component).toMatchSnapshot();
  });

  it("Render component correctly with props", () => {
    const component = mount(<Post post={ARTICLES[0]} />);
    expect(component.find("p").text()).toBe("90 views");
    expect(component.find(Link).length).toBe(1);
    expect(component.find(Link).prop("href")).toBe("/post/1");
    expect(component.find(Image).prop("src")).toBe(
      "https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg"
    );
  });
});
