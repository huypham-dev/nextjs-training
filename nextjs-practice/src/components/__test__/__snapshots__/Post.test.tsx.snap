// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Snapshot 1`] = `
<div
  className="w-full md:w-1/2 lg:w-1/3 px-6 p-5"
>
  <div
    className="bg- p-4 rounded-md shadow-xl border-gray-200 border-2"
  >
    <Memo(Typography)
      className="text-md font-extrabold tracking-widest mb-2"
      variant="h2"
    >
      Lorem ipsum dolor sit amet, consectetur adipisicing elit
    </Memo(Typography)>
    <div
      className="w-full text-center"
    >
      <Image
        alt="article image"
        className="rounded-lg"
        height={150}
        layout="responsive"
        src="https://genk.mediacdn.vn/139269124445442048/2021/3/2/photo-1-16146533712881100681556.jpg"
        width={250}
      />
    </div>
    <Memo(Typography)
      className="text-xs text-gray-500 mb-4"
    >
      90
       
      <i>
        views
      </i>
    </Memo(Typography)>
    <Memo(LinkBase)
      className="inline-block border-2 border-black py-3 px-8 uppercase font-display font-thin tracking-wide text-xs hover:bg-yellow-400 hover:border-yellow-400"
      href="/post/1"
    >
      Read more
    </Memo(LinkBase)>
  </div>
</div>
`;
